### PR TITLE
core[patch]: Fix RemoteRunnable streaming when used in sequences, add tracing

### LINKED ISSUE #4811: Streaming does not work if RemoteRunnable is part of RunnableSequence
### Checked other resources

- [X] I added a very descriptive title to this issue.
- [X] I searched the LangChain.js documentation with the integrated search.
- [X] I used the GitHub search to find a similar question and didn't find it.
- [X] I am sure that this is a bug in LangChain.js rather than my code.
- [X] The bug is not resolved by updating to the latest stable version of LangChain (or the specific integration package).

#### Langserve endpoint

```python
from fastapi import FastAPI
from langserve import add_routes

app = FastAPI()
app.add_middleware(
    CORSMiddleware,
    allow_origins=["*"],
    allow_credentials=True,
    allow_methods=["*"],
    allow_headers=["*"],
    expose_headers=["*"],
)

OPENAI_API_KEY= "your API key"

add_routes(app, ChatOpenAI(model="gpt-3.5-turbo-1106", api_key=OPENAI_API_KEY), path='/models/ChatOpenAI')

if __name__ == "__main__":
    import os
    import uvicorn

    HTTP_PORT = 8081
    uvicorn.run(app, host="0.0.0.0", port=HTTP_PORT)
```

#### Python's RemoteRunnable streams correctly

```python
from langserve.client import RemoteRunnable;
from langchain_core.prompts import ChatPromptTemplate

prompt = ChatPromptTemplate.from_template("Tell me a joke about {topic}")
model = RemoteRunnable("http://localhost:8081/models/ChatOpenAI")
chain = prompt.pipe(model)

stream = chain.stream({'topic': 'cats'})
for chunk in stream:
    print(chunk)
```

#### JS equivalent does not stream and just calls invoke

```typescript
import { ChatPromptTemplate } from '@langchain/core/prompts';
import { RemoteRunnable } from "langchain/runnables/remote";

const prompt = ChatPromptTemplate.fromTemplate("Tell me a joke about {topic}")
const model = new RemoteRunnable({ url: `http://localhost:8081/models/ChatOpenAI` });
const chain = RunnableSequence.from([prompt, model]);

const stream = chain.stream({topic: "cats"});

for await (const chunk of stream) {
  console.log(chunk);
}
```

### Solo RemoteRunnable in JS streams properly

When we simplify the chain and just call `.stream` directly on RemoteRunnable, streaming seems to be working correctly:

```typescript
import { RemoteRunnable } from "langchain/runnables/remote";

const chain = new RemoteRunnable({ url: `http://localhost:8081/models/ChatOpenAI` });

const stream = chain.stream({topic: "cats"});

for await (const chunk of stream) {
  console.log(chunk);
}
```

### Error Message and Stack Trace (if applicable)

_No response_

### Description

I'm trying to use RemoteRunnable in place of a model in my javascript LCEL chains but when i compose the RemoteRunnable with any other runnable to form a RunnableSequence, it does not stream and fallbacks to invoke.

Equivalent code works in python version.

### pnpm show langchain 

```
❯ pnpm show langchain

langchain@0.1.28 | MIT | deps: 18 | versions: 259
Typescript bindings for langchain
https://github.com/langchain-ai/langchainjs/tree/main/langchain/

keywords: llm, ai, gpt3, chain, prompt, prompt engineering, chatgpt, machine learning, ml, openai, embeddings, vectorstores

dist
.tarball: https://registry.npmjs.org/langchain/-/langchain-0.1.28.tgz
.shasum: fd07a9880a68865e5c803dc01decdaa097b7c756
.integrity: sha512-LNuILtOSFptfOqDwUwQAl7B0dYrBtxzVZ6lDhc5DA8rk+pyTCy12HQO7t8bGqC8oQlEPYPKaom8VRlvfupUBTw==
.unpackedSize: 4.1 MB

dependencies:
@anthropic-ai/sdk: ^0.9.1     expr-eval: ^2.0.2             langsmith: ~0.1.7             yaml: ^2.2.1
@langchain/community: ~0.0.36 js-tiktoken: ^1.0.7           ml-distance: ^4.0.0           zod-to-json-schema: ^3.22.3
@langchain/core: ~0.1.44      js-yaml: ^4.1.0               openapi-types: ^12.1.3        zod: ^3.22.4
@langchain/openai: ~0.0.19    jsonpointer: ^5.0.1           p-retry: 4
binary-extensions: ^2.2.0     langchainhub: ~0.0.8          uuid: ^9.0.0

maintainers:
- nfcampos <nuno@boringbits.io>
- jacoblee93 <jacoblee93@gmail.com>
- davidduong <david@duong.cz>
- hwchase17 <hw.chase.17@gmail.com>
- basproul <braceasproul@gmail.com>

dist-tags:
latest: 0.1.28     next: 0.1.29-rc.0

published 3 days ago by jacoblee93 <jacoblee93@gmail.com>
```

### Node version

```
❯ node --version
v21.5.0
```

### pnpm version

```
❯ pnpm --version
8.14.1
```

### PR DESCRIPTION
**Pull request by rahilvora**
This PR is trying to solve issue where `RemoteRunnable` object does not support streaming with `RunnableSequence`

`Code Flow`
`RunnableSequence` and `RemoteRunnable` both the classes extends `Runnable` class

`Runnable` class has implementation of `stream`. Now, `const chain = RunnableSequence.from([prompt, model])`  returns `RunnableSequence`  object. `chain.stream` would call stream function on `Runnable` class (parent class) as `RunnableSequence`  class (child class) does not have stream function on it.

Now, `stream` function in `Runnable` class calls `this._streamIterator` . In this context, `this` is `RunnableSequence` , Hence it will call `_streamIterator` of `RunnableSequence`.

`_streamIterator` calls `object.transform` function, which calls `transform` function of `Runnable` class. `transform` function internally calls `this._streamIterator` again. Now, `this` is `object context` which is `RemoteRunnable`

But, in case of `RemoteRunnable` object (which is passed as model in above example), does not have implementation of `_streamIterator` function hence it calls `_streamIterator` of parent class i.e. `Runnable class` which internally is yielding `invoke` `(this.invoke(input, options);`. Hence, this calls `RemoteRunnable invoke function`.

`Fix`

We need to override `_streamIterator` for RemoteRunnable class, so that it yield `stream` instead of `invoke`.

Fixes #4811

### CODE REVIEW

**Comment by jacoblee93** [comment]
Thanks for this!

Would be good to add a test confirming the fix?

File: langchain-core/src/runnables/tests/runnable_remote.test.ts [hunk 1]
```diff
@@ -215,4 +215,20 @@ describe("RemoteRunnable", () => {
       AIMessage
     );
   });
+
+  test("StreamIterator yield stream model output", async () => {
+    const remote = new RemoteRunnable({ url: `${BASE_URL}/b` });
+    const streamIterator = await remote._streamIterator({ text: "What are the 5 best apples?" });
```
Comment (jacoblee93) [inline langchain-core/src/runnables/tests/runnable_remote.test.ts:221]: This should test streaming usage in a chain though?

`const chain = prompt.pipe(remote).stream();`

Comment (rahilvora) [reply langchain-core/src/runnables/tests/runnable_remote.test.ts:221]: I thought I would just test `_streamIterator` function which I added. But, I think what you said make sense. Let me try with chain.

Comment (rahilvora) [reply langchain-core/src/runnables/tests/runnable_remote.test.ts]: Updated!!

**Comment by jacoblee93** [comment]
Thanks!

**Pull request merged**